akka-http/src/main/scala/akka/http/scaladsl/server/directives/CodingDirectives.scala

```diff
@@ -82,14 +82,20 @@ trait CodingDirectives {
           val effectiveDecoder = decoder.withMaxBytesPerChunk(settings.decodeMaxBytesPerChunk)
           mapRequest { request ⇒
             effectiveDecoder.decodeMessage(request).mapEntity { entity ⇒
-              entity.transformDataBytes(Flow[ByteString].recover {
-                case NonFatal(e) ⇒
-                  throw IllegalRequestException(
-                    StatusCodes.BadRequest,
-                    ErrorInfo("The request's encoding is corrupt", e.getMessage))
-              })
+              {
+                val recovered = entity.transformDataBytes(Flow[ByteString].recover {
+                  case NonFatal(e) ⇒
+                    throw IllegalRequestException(
+                      StatusCodes.BadRequest,
+                      ErrorInfo("The request's encoding is corrupt", e.getMessage))
+                })
+                recovered match {
+                  case c: HttpEntity.Chunked ⇒ c.copy(chunks = HttpEntity.limitableChunkSource(c.chunks))
+                  case e: HttpEntity         ⇒ e
+                }
+              }
             }
-          }
+          } & withSizeLimit(settings.decodeMaxSize)
```

Comment:
Ok, so this together with the limitableChunkSource is the fix right? My brain is tired: Is it only ever streamed for chunked, not a large non-chunked entity?

Reply:
A large non-chunked entity should still be restricted by `withSizeLimit` - but I'll make sure the tests cover both cases.

Reply:
The output of `transformDataBytes` will be a chunked entity (but maybe it's still a brittle assumption). I think we should change `tranformDataBytes`, instead, to ensure that the result is always limitable. WDYT?

Reply:
Ultimately, it's the `HttpEntity.withSizeLimit` that is broken right now (it's debatable whether it was a good idea to add it there but we cannot change that right now :) ).

What if we would always add the limitable stage when `withSizeLimit` is called (at least if we cannot verify whether the last stage in that source already is a Limitable which isn't possible using public streams API afaik). That might mean that we have multiple such stages in the source e.g. in the default case with `decodeRequest` after this PR. For a slight performance overhead it would still be the safer option.

The performance overhead would be less than currently if no limits are defined in the first place because then we would never add the stage.

Reply:
Hmm that sounds interesting, tried that! I'm not too familiar with how exactly Attributes work, so I'd appreciate if you could double-check I'm not introducing any problems there.